extensions/arc/src/ui/tree/controllerTreeNode.ts

```diff
@@ -13,26 +13,40 @@ import { ControllerDashboard } from '../dashboards/controller/controllerDashboar
 import { PostgresModel } from '../../models/postgresModel';
 import { parseInstanceName } from '../../common/utils';
 import { MiaaModel } from '../../models/miaaModel';
+import { Deferred } from '../../common/promise';
+import { RefreshTreeNode } from './refreshTreeNode';
 
 /**
  * The TreeNode for displaying an Azure Arc Controller
  */
 export class ControllerTreeNode extends TreeNode {
 
 	private _children: TreeNode[] = [];
+	private _childrenRefreshPromise = new Deferred();
 
-	constructor(private _model: ControllerModel, private _context: vscode.ExtensionContext) {
-		super(_model.controllerUrl, vscode.TreeItemCollapsibleState.Collapsed, ResourceType.dataControllers);
-		_model.onRegistrationsUpdated(registrations => this.refreshChildren(registrations));
-		_model.refresh().catch(err => console.log(`Error refreshing Arc Controller model for tree node : ${err}`));
+	constructor(public model: ControllerModel, private _context: vscode.ExtensionContext) {
+		super(model.info.url, vscode.TreeItemCollapsibleState.Collapsed, ResourceType.dataControllers);
+		model.onRegistrationsUpdated(registrations => this.refreshChildren(registrations));
 	}
 
 	public async getChildren(): Promise<TreeNode[]> {
+		// First reset our deferred promise so we're sure we'll get the refreshed children
+		this._childrenRefreshPromise = new Deferred();
+		try {
+			await this.model.refresh();
+			await this._childrenRefreshPromise.promise;
```

Comment:
I'm sure you know what you're doing, but I can't figure out how this deferred promise works.  Presumably this awaits until refreshChildren resolves the deferred promise, but then who calls refreshChildren?

Reply:
That's from the event callback for the model for when the registrations are updated. So the flow looks like this :

1. Model refresh is called
1. Then this code sits there waiting (because we need the children in order to continue)
1. Once the model is finished refreshing the children it fires the `onRegistrationsUpdated` event which calls `this.refreshChildren` to update the local cache of children nodes and resolves the deferred promise
1. Now the getChildren call continues and returns the list of children that were retrieved

Reply:
Here's the event callback - https://github.com/microsoft/azuredatastudio/pull/11014/files#diff-7681f15b376515153dfce0fd23d318f0R29

Reply:
That makes total sense now, thanks.